README.rst
===================
requests_downloader
===================


.. image:: https://img.shields.io/pypi/v/requests_downloader?color=success
        :target: https://pypi.python.org/pypi/requests_downloader

.. image:: https://readthedocs.org/projects/requests_downloader/badge/?version=latest
        :target: https://requests_downloader.readthedocs.io/en/latest/?version=latest
        :alt: Documentation Status

.. image:: https://img.shields.io/pypi/pyversions/requests_downloader
        :target: https://pypi.python.org/pypi/requests_downloader
        :alt: Python Version Support

.. image:: https://img.shields.io/github/issues/hrishikeshrt/requests_downloader
        :target: https://github.com/hrishikeshrt/requests_downloader/issues
        :alt: GitHub Issues

.. image:: https://img.shields.io/github/followers/hrishikeshrt?style=social
        :target: https://github.com/hrishikeshrt
        :alt: GitHub Followers

.. image:: https://img.shields.io/twitter/follow/hrishikeshrt?style=social
        :target: https://twitter.com/hrishikeshrt
        :alt: Twitter Followers


Python package to download files


* Free software: GNU General Public License v3
* Documentation: https://requests_downloader.readthedocs.io.


Features
========

* Hassle-free download using ``requests``
* Download from Drive, Dropbox, Archive or direct URLs
* No need to specify a name for the file to be downloaded
* Command Line Interface to download
* External ``requests.Session`` object can be passed

Usage
=====

Use in a Project
----------------

Get multiple download options (e.g. for ``archive.org`` links):

.. code-block:: python

    from requests_downloader import downloader
    download_urls, default_idx = downloader.handle_url('<url>')


Download a file:

.. code-block:: python

    from requests_downloader import downloader
    downloader.download('<download_url>')

Use Console Interface
---------------------

.. code-block:: console

    usage: smart-dl [-h] [--download_dir DOWNLOAD_DIR] [--download_file DOWNLOAD_FILE]
                    [--download_path DOWNLOAD_PATH] [--block BLOCK] [--timeout TIMEOUT]
                    [--resume] [--progress] [--checksum CHECKSUM] [--verbose] [--debug]
                    [--version] url

    positional arguments:
    url                   Download URL

    optional arguments:
    -h, --help            show this help message and exit
    --download_dir DOWNLOAD_DIR
                            Specify downloads directory
    --download_file DOWNLOAD_FILE
                            Specify filename
    --download_path DOWNLOAD_PATH
                            Specify path (ignores _dir or _file arguments)
    --block BLOCK         Block size while writing the file, in bytes
    --timeout TIMEOUT     Timeout in seconds
    --resume              Try to resume the download, if supported
    --progress            Show download progressbar
    --checksum CHECKSUM   Checksum to verify integrity of the download
    --verbose             Enable verbose output
    --debug               Enable debug information
    --version             show program's version number and exit


Credits
=======

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
History
=======

0.4.0 (2022-04-28)
------------------

* Option to show filename to the left of progressbar
* Modularize code
* Style: Black
* Fix bugs

0.3.0 (2022-02-11)
------------------

* Several documentation updates
* Code refactoring
* Change console command from clunky `requests_downloader` to `smart-dl`

0.2.0 (2021-06-08)
------------------

* Improvements in content disposition and content length handling
* Logging improvements

0.1.7 (2021-06-07)
------------------

* Minor bugfixes and spelling corrections.

0.1.5 (2020-11-03)
------------------

* Support for Google Docs (Document, Spreadsheets, Presentations).

0.1.4 (2020-09-07)
------------------

* Bugfixes.
* First Alpha release.

0.1.0 (2020-09-05)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for requests_downloader."""


tests/test_utils.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Tests for `requests_downloader.utils`."""

from requests_downloader.utils import md5sum

###############################################################################


def test_md5sum(tmp_path):
    tmp_file = tmp_path / "test.txt"
    tmp_file.write_text("hello world\n")
    assert "6f5902ac237024bdd0c176cb93063dc4" == md5sum(tmp_file)


tests/test_handlers.py
#!/usr/bin/env python
"""Tests for `requests_downloader.handlers`."""

from requests_downloader.handlers import handle_url


def test_handle_url():
    url_result = {
        'https://drive.google.com/file/d/<file_id>/view': [
            ('drive', 'https://drive.google.com/u/0/uc?id=<file_id>&export=download')
        ],
        'https://drive.google.com/file/d/<file_id>/view?usp=sharing': [
            ('drive', 'https://drive.google.com/u/0/uc?id=<file_id>&export=download')
        ],
        'https://www.dropbox.com/s/<file_id>/<file_name>': [
            ('dropbox', 'https://www.dropbox.com/s/<file_id>/<file_name>?dl=1')
        ],
        'https://www.dropbox.com/s/<file_id>/<file_name>?dl=0': [
            ('dropbox', 'https://www.dropbox.com/s/<file_id>/<file_name>?dl=1')
        ],
        'https://www.dropbox.com/s/<file_id>/<file_name>?dl=0&opt=val': [
            ('dropbox', 'https://www.dropbox.com/s/<file_id>/<file_name>?dl=1&opt=val')
        ],
    }
    for url, url_result in url_result.items():
        assert handle_url(url) == (url_result, 0)


requests_downloader/handlers.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Special URL Handlers
"""


import re
import logging
from urllib.parse import urlparse, urlunparse

import requests
from bs4 import BeautifulSoup

###############################################################################

LOGGER = logging.getLogger(__name__)

###############################################################################


def handle_url(url):
    """
    Infer the actual download URLs by handling various special cases

    Parameters
    ----------
    url : str
        Provided download URL.

    Returns
    -------
    urls : list
        List of inferred download URLs from the provided URL.
    default_idx: int
        Index of default URL to download.
    """
    default_idx = 0
    drive = "https://drive.google.com"
    drive_pattern_1 = rf"{drive}/file/d/([^\/]*)/.*"
    drive_match_1 = re.match(drive_pattern_1, url)
    if drive_match_1:
        LOGGER.debug("Google Drive pattern-1 matched.")
        file_id = drive_match_1.group(1)
        dl_url = f"{drive}/u/0/uc?id={file_id}&export=download"
        return [("drive", dl_url)], default_idx

    drive_pattern_2 = rf"{drive}/open\?id=([^\/&]*).*"
    drive_match_2 = re.match(drive_pattern_2, url)
    if drive_match_2:
        LOGGER.debug("Google Drive pattern-2 matched.")
        file_id = drive_match_2.group(1)
        dl_url = f"{drive}/u/0/uc?id={file_id}&export=download"
        return [("drive", dl_url)], default_idx

    docs = "https://docs.google.com"
    preference = {
        "spreadsheets": ["xlsx", "ods", "pdf"],
        "document": ["docx", "odt", "pdf"],
        "presentation": ["pptx", "odp", "pdf"],
    }
    doc_pattern = rf"{docs}/(spreadsheets|document|presentation)/d/([^\/]*)/.*"
    doc_match = re.match(doc_pattern, url)
    if doc_match:
        LOGGER.debug("Google Docs pattern matched.")
        doc_type = doc_match.group(1)
        doc_id = doc_match.group(2)
        LOGGER.debug(f"Type: {doc_type}, ID: {doc_id}")
        dl_types = preference[doc_type]
        dl_urls = [
            (
                dl_type,
                (
                    f"{docs}/{doc_type}/d/{doc_id}/export/{dl_type}"
                    f"?id={doc_id}"
                ),
            )
            if doc_type == "presentation"
            else (
                dl_type,
                (
                    f"{docs}/{doc_type}/d/{doc_id}/export?"
                    f"format={dl_type}&id={doc_id}"
                ),
            )
            for dl_type in dl_types
        ]
        return dl_urls, default_idx

    archive = "https://archive.org"
    archive_pattern = rf"{archive}/(details|download)/([^\/]*).*"
    archive_match = re.match(archive_pattern, url)
    if archive_match:
        LOGGER.debug("Archive.org pattern matched.")
        content_name = archive_match.group(2)
        archive_url = f"{archive}/download/{content_name}"
        dl_url = f"{archive}/compress/{content_name}"
        dl_urls = [("all", dl_url)]
        default_idx = 0

        r = requests.get(archive_url)
        soup = BeautifulSoup(r.content, "html.parser")
        div = soup.find("div", class_="download-directory-listing")
        urls = div.find_all("a")
        dl_urls += [
            (
                url["href"].split(".")[-1],
                f"{archive}/download/{content_name}/{url['href']}",
            )
            for url in urls
            if (
                not url["href"].endswith(content_name)
                and not url["href"].endswith("/")
                and url.get_text().find("View Contents") == -1
            )
        ]
        preference_order = ["pdf", "mp3", "all"]
        for preference in preference_order:
            for idx, (tag, url) in enumerate(dl_urls):
                if tag == preference:
                    default_idx = idx
                    break
            else:
                continue
            break

        return dl_urls, default_idx

    dropbox_pattern = "dropbox.com"
    dropbox_match = dropbox_pattern in url
    if dropbox_match:
        parse_result = urlparse(url)
        query = dict(
            p.split("=") for p in parse_result.query.split("&") if "=" in p
        )
        query["dl"] = 1
        query_string = "&".join([f"{k}={v}" for k, v in query.items()])
        parse_result = parse_result._replace(query=query_string)
        return [("dropbox", urlunparse(parse_result))], default_idx

    LOGGER.debug("No specific pattern matched.")
    return [("direct", url)], default_idx


requests_downloader/downloader.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Main module containing download function
"""

###############################################################################

import os
import logging
import mimetypes
from urllib.parse import unquote

import requests
from tqdm import tqdm

from .handlers import handle_url
from .utils import md5sum

###############################################################################

LOGGER = logging.getLogger(__name__)

###############################################################################

HEADERS = {
    "Range": "bytes=0-",
    "User-Agent": (
        "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:89.0) "
        "Gecko/20100101 Firefox/89.0"
    ),
    "Upgrade-Insecure-Requests": "1",
    "Connection": "keep-alive",
    "Keep-Alive": "timeout=10, max=100",
}

###############################################################################


def download(
    url,
    download_dir="",
    download_file=None,
    download_path=None,
    headers={},
    session=None,
    block_size=1024,
    timeout=60,
    resume=True,
    show_progress=True,
    show_progress_desc=True,
    max_desc_length=35,
    checksum=None,
    smart=True,
    url_handler=None,
):
    """
    Download a file

    Parameters
    ----------
    url : str
        URL to download.
    download_dir : str, optional
        Path of the directory to download the file in.
        The default is '' (i.e. current directory).
    download_file : str, optional
        Name for the downloaded file.
        If None, the function will infer it from URL and Content-Disposition
        The default is None.
    download_path : str, optional
        Full path where the downloaded file should be saved.
        If None, the function will save it in `download_dir/download_file`
        If provided, `download_dir` and `download_file` arguments are ignored.
        The default is None.
    headers : dict, optional
        Headers to be sent.
        The default is {}.
        Note:
            * These headers are merged with a default set of headers.
            * In case of a conflict the user-provided values are used.
            * This behaviour is inherited from `requests.Session()`
    session : object, optional
        A valid `requests.Session` object.
        This is useful when download url requires authentication.
        In such a case, authentication can be handled independently in session.
        The default is None.
    block_size : int, optional
        Block size, in bytes, to stream the downloadable content.
        The default is 1024.
    timeout : float, optional
        Timeout, in seconds
        The default is 60.
    resume : bool, optional
        Try to resume download.
        The default is True.
    show_progress : bool, optional
        Show progressbar.
        The default is True.
    show_progress_desc : str or bool, optional
        Show the description to the left of progressbar.
        If False or None, no description is shown.
        If True, the name of file being downloaded is shown.
        Otherwise, the `str()` of the provided value is shown.
        The default is True.
    max_desc_length : int, optional
        If length of the description is more, abbreviate it by showing the
        first and last parts connected by three dots.
        The default is 35.
    checksum : str, optional
        Value of md5 checksum of the file to be downloaded.
        If provided, the downloaded file will be verified using the checksum.
        The default is None.
    smart : bool, optional
        Use url_handler for special case URLs
        The default is True.
    url_handler : function, optional
        Handler function for special cases of download URLs
        The function should return a list of (TAG, URL) pairs and default index

    Returns
    -------
    download_path: str or None
        If download was successful, full `download_path`
        otherwise, None
    """
    success = True
    if smart:
        if url_handler is None:
            url_handler = handle_url
        urls, url_idx = url_handler(url)
        url = urls[url_idx][1]

    LOGGER.debug(f"URL: {url}")

    if session is None:
        session = requests.Session()
        session.headers.update(HEADERS)

    LOGGER.debug(session.headers)
    r = session.head(url, headers=headers, timeout=timeout)

    resume_supported = r.headers.get("accept-ranges") == "bytes"
    file_mode = "ab" if resume_supported else "wb"
    LOGGER.debug(f"Resume Supported: {resume_supported}")

    r = session.get(url, headers=headers, timeout=timeout, stream=True)
    LOGGER.debug(r.headers)

    content_length = int(r.headers.get("content-length", 0))
    LOGGER.debug(f"Content-Length: {content_length}")

    content_range = r.headers.get("content-range", "")
    _content_range_part = content_range.split("/")[-1].strip()
    LOGGER.debug(f"Content-Range: {content_range}")

    if content_length == 0 and _content_range_part:
        content_length = int(_content_range_part)
        LOGGER.debug(f"Content-Length (from Range): {content_length}")

    content_type = r.headers.get("content-type")
    html_content = content_type == "text/html; charset=utf-8"
    LOGGER.debug(f"Content-Type: {content_type}")
    LOGGER.debug(f"HTML Content: {html_content}")

    if html_content:
        LOGGER.error("HTML content detected.")
        LOGGER.error(f"Download from {url} aborted.")
        return False

    extension_guess = mimetypes.guess_extension(content_type)
    LOGGER.debug(f"Extension Guess: {extension_guess}")

    visible_name = r.url.split("/")[-1]
    if extension_guess and not visible_name.endswith(extension_guess):
        visible_name += f".{extension_guess}"
    visible_name = unquote(visible_name, "UTF-8")
    LOGGER.debug(f"Visible Name: {visible_name}")

    provided_name = None
    cd = r.headers.get("content-disposition", None)
    LOGGER.debug(f"Content-Disposition: {cd}")
    if cd is not None:
        cd_fields = {}
        for part in cd.split(";"):
            _kv = part.split("=")
            if "=" in part:
                _k = _kv[0].strip(" \t\n\"'")
                _v = _kv[1].strip(" \t\n\"'")
                cd_fields[_k] = _v

        LOGGER.debug(cd_fields)
        provided_names = [v for k, v in cd_fields.items() if k == "filename"]
        provided_encoded_names = [
            v for k, v in cd_fields.items() if k == "filename*"
        ]

        # provided_names = re.findall('filename="(.+)"', cd)
        LOGGER.debug(f"Filenames: {provided_names}")
        LOGGER.debug(f"Filenames*: {provided_encoded_names}")

        if provided_names:
            provided_name = provided_names[0]

        if provided_encoded_names:
            encoding, name = provided_encoded_names[0].split("''")
            LOGGER.debug(f"Encoding: '{encoding}', Name: '{name}'")
            provided_name = unquote(name, encoding=encoding)

        LOGGER.debug(f"Final Provided Name: {provided_name}")

    if not download_file:
        download_file = provided_name if provided_name else visible_name

    if not download_file:
        LOGGER.error("Download location could not be inferred.")
        LOGGER.error(f"Download from {url} aborted.")
        return False

    if not download_path:
        download_path = os.path.join(download_dir, download_file)
    else:
        download_file = os.path.basename(download_path)

    LOGGER.info(
        f"Downloading '{download_file}' ... " f"({content_length} bytes)"
    )

    with open(download_path, "ab") as f:
        position = f.tell()
        LOGGER.debug(f"Current Position: {position}")
        if position and position == content_length:
            LOGGER.info(f"File '{download_file}' is already downloaded!")
            return download_path

    wrote = 0
    with open(download_path, file_mode) as f:
        position = f.tell()
        if resume and resume_supported:
            if position:
                headers["Range"] = f"bytes={position}-"
                LOGGER.info(
                    f"Resuming '{download_file}' from {position} bytes"
                )

            r = session.get(url, headers=headers, timeout=timeout, stream=True)

        if show_progress_desc:
            if show_progress_desc is True:
                desc = download_file
            else:
                desc = str(show_progress_desc)
            if len(desc) > max_desc_length:
                prefix_length = (max_desc_length - 3) // 2
                suffix_length = prefix_length
                desc = f"{desc[:prefix_length]}...{desc[-suffix_length:]}"
        else:
            desc = None
        with tqdm(
            initial=position,
            desc=desc,
            total=content_length,
            unit="B",
            unit_scale=True,
            disable=not show_progress,
        ) as t:
            for data in r.iter_content(block_size):
                wrote += f.write(data)
                t.update(len(data))

    LOGGER.debug(f"Wrote: {wrote}")

    if content_length == 0:
        filesize = os.stat(download_path).st_size
        LOGGER.debug(f"Filesize: {filesize}")
        if not filesize:
            os.unlink(download_path)
            LOGGER.warning(
                f"Downloaded file '{download_file}' was empty and was removed."
            )
            success = False
        else:
            LOGGER.warning(
                f"Integrity of '{download_file}' could not verified."
            )
    elif (position + wrote) != content_length:
        success = False
        LOGGER.warning(f"Inconsistency in download from '{url}'.")
        LOGGER.debug(
            f"Wrote {wrote} bytes out of {content_length - position}."
        )

    if checksum is not None:
        download_checksum = md5sum(download_path)
        if download_checksum != checksum:
            success = False
            LOGGER.warning("Invalid checksum.")
            LOGGER.debug(
                f"md5sum({download_file}) = {download_checksum} != {checksum})"
            )

    if success:
        LOGGER.info(f"Successfully downloaded '{download_file}' from '{url}'.")
        return download_path
    else:
        LOGGER.info(
            f"An error occurred in downloading '{download_file}' from '{url}'."
        )
        return False


###############################################################################


requests_downloader/cli.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Command Line Interface for requests_downloader
"""

###############################################################################

import sys
import logging
import argparse

from . import __version__
from .downloader import download
from .handlers import handle_url

###############################################################################

ROOT_LOGGER = logging.getLogger()
if not ROOT_LOGGER.hasHandlers():
    ROOT_LOGGER.addHandler(logging.StreamHandler())
ROOT_LOGGER.setLevel(logging.WARNING)

###############################################################################


def main():
    """CLI for requests_downloader"""
    parser = argparse.ArgumentParser()
    parser.add_argument("url", help="Download URL")
    parser.add_argument(
        "--download_dir", help="Specify downloads directory", default=""
    )
    parser.add_argument(
        "--download_file", help="Specify filename", default=None
    )
    parser.add_argument(
        "--download_path",
        help="Specify path (ignores _dir or _file arguments)",
        default=None,
    )
    parser.add_argument(
        "--block",
        help="Block size while writing the file, in bytes",
        default=1024,
    )
    parser.add_argument("--timeout", help="Timeout in seconds", default=60)
    parser.add_argument(
        "--resume",
        help="Try to resume the download, if supported",
        action="store_false",
    )
    parser.add_argument(
        "--progress", help="Show download progressbar", action="store_false"
    )
    parser.add_argument(
        "--checksum",
        help="Checksum to verify integrity of the download",
        default=None,
    )
    parser.add_argument(
        "--verbose", help="Enable verbose output", action="store_true"
    )
    parser.add_argument(
        "--debug", help="Enable debug information", action="store_true"
    )
    parser.add_argument(
        "--version", action="version", version="%(prog)s " + __version__
    )
    args = vars(parser.parse_args())

    if args["verbose"]:
        ROOT_LOGGER.setLevel(logging.INFO)
    if args["debug"]:
        ROOT_LOGGER.setLevel(logging.DEBUG)

    urls, url_idx = handle_url(args["url"])
    if len(urls) > 1:
        options = [
            f"{idx+1:>3}. | {s.upper():<8}| {u}"
            for idx, (s, u) in enumerate(urls)
        ]
        print("\n".join(options))
        valid_responses = [str(i + 1) for i in range(len(options))]
        while True:
            response = input(
                f"Please choose a file to download (default: {url_idx + 1}): "
            )
            if not response:
                response = url_idx
                break

            if response not in valid_responses:
                print(f"Please enter a value between 1 to {len(options)}.")
                continue
            else:
                response = int(response) - 1
                break
    else:
        response = url_idx

    url = urls[response][1]
    location = download(
        url,
        download_dir=args["download_dir"],
        download_file=args["download_file"],
        download_path=args["download_path"],
        block_size=int(args["block"]),
        timeout=float(args["timeout"]),
        resume=args["resume"],
        show_progress=args["progress"],
        smart=False,
        checksum=args["checksum"],
    )
    print(f"File saved to '{location}'.")

    return 0


###############################################################################


if __name__ == "__main__":
    sys.exit(main())
